cmd/veneur-emit/main.go

```diff
@@ -0,0 +1,135 @@
+	name       = flag.String("name", "", "Name of metric to report. Ex: daemontools.service.starts")
+	gauge      = flag.Float64("gauge", 0, "Report a 'gauge' metric. Value must be float64.")
+	timing     = flag.Duration("timing", time.Now().Sub(time.Now()), "Report a 'timing' metric. Value must be parseable by time.ParseDuration.")
+	timeinms   = flag.Float64("timeinms", 0, "Report a 'timing' metric, in milliseconds. Value must be float64.")
+	count      = flag.Int64("count", 0, "Report a 'count' metric. Value must be an integer.")
+	tag        = flag.String("tag", "", "Tag(s) for metric, comma separated. Ex: service:airflow")
+	debug      = flag.Bool("debug", false, "Turns on debug messages.")
+)
+
+// MinimalClient represents the functions that we call on Clients in veneur-emit.
+type MinimalClient interface {
+	Gauge(name string, value float64, tags []string, rate float64) error
+	Count(name string, value int64, tags []string, rate float64) error
+	Timing(name string, value time.Duration, tags []string, rate float64) error
+	TimeInMilliseconds(name string, value float64, tags []string, rate float64) error
+}
+
+func main() {
+	passedFlags := flags()
+
+	if *debug {
+		logrus.SetLevel(logrus.DebugLevel)
+	}
+
+	var config *veneur.Config
+	if passedFlags["f"] {
+		conf, err := veneur.ReadConfig(*configFile)
+		if err != nil {
+			logrus.WithError(err).Fatal("Error reading configuration file.")
+		}
+		config = &conf
+	}
+
+	addr, err := addr(passedFlags, config, hostport)
+	if err != nil {
+		logrus.WithError(err).Fatal("Error getting destination address.")
+	}
+	logrus.Debugf("destination: %s", addr)
+
+	var conn MinimalClient
+	conn, err = statsd.New(addr)
+	if err != nil {
+		logrus.WithError(err).Fatal("Error!")
+	}
+
+	tags := tags(*tag)
+	err = sendMetrics(conn, passedFlags, *name, tags)
+	if err != nil {
+		logrus.WithError(err).Fatal("Error sending metric(s).")
+	}
+}
+
+func flags() map[string]bool {
+	flag.Parse()
+	// hacky way to detect which flags were *actually* set
+	passedFlags := make(map[string]bool)
+	flag.Visit(func(f *flag.Flag) {
+		passedFlags[f.Name] = true
+	})
+	return passedFlags
+}
+
+func tags(tag string) []string {
+	var tags []string
+	for _, elem := range strings.Split(tag, ",") {
+		tags = append(tags, elem)
+	}
+	return tags
+}
+
+func addr(passedFlags map[string]bool, conf *veneur.Config, hostport *string) (string, error) {
+	addr := ""
+	var err error
+	if passedFlags["f"] && (conf != nil) {
+		addr = conf.UdpAddress // TODO: conf.TcpAddress
+	} else if passedFlags["hostport"] && !(hostport == nil || *hostport == "" || !strings.Contains(*hostport, ":")) {
+		addr = *hostport
+	} else {
+		err = errors.New("you must either specify a Veneur config file or a valid hostport")
+	}
+	return addr, err
+}
+
+func sendMetrics(client MinimalClient, passedFlags map[string]bool, name string, tags []string) error {
+	var err error
+	if passedFlags["gauge"] {
```

Comment:
I don't think it's really the right time to do this, but I think this could be improved by creating a metric object and passing it in to this method, such that it doesn't have to do the parsing/construction. Only the serialization. So I'd say that when we get around to doing the SSF bits we can make that happen.